From faee6f1bedad9eb5a5047718b1210377ce71829f Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 16 May 2022 20:55:03 +0300
Subject: [PATCH] [SW-86866] gaudi3: update udelay routine to use psoc ts

Currently, udelay logic make use of fixed cpu frequency and runs nops
based on the delays requested. However there is possibility that
ARC may run in different frequencies and will impact delay behaviour.
This patch changes udelay logic to make use of psoc timestamp for
delay calculation.

Change-Id: Id8d88d32439d47088db1b1ba62d6dcd880b26ca5
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index de72709..57efed0 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -368,6 +368,8 @@
 
 	arcboot_rom_status_update(PROGRESS_INIT);
 
+	arcboot_rom_init_ts();
+
 	arcboot_rom_prstn_init();
 
 	arcboot_rom_read_cold_flops();
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 582c5d2..414a383 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -209,8 +209,6 @@
 
 	device_init();
 
-	arcboot_rom_init_ts();
-
 	security_rr_recovery_config();
 
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
diff --git a/common/delay.c b/common/delay.c
index 934049b..01ff97c 100644
--- a/common/delay.c
+++ b/common/delay.c
@@ -67,3 +67,27 @@
 
 	return false;
 }
+
+void udelay(unsigned int usec)
+{
+	uint64_t start, current, end = cycles_per_usec * usec;
+	bool overlap = false;
+
+	start = udelay_get_ts();
+	end += start;
+
+	if (end < start)
+		overlap = true;
+
+	while (1) {
+		current = udelay_get_ts();
+		if (!overlap) {
+			/* Check also for unplanned overlap */
+			if (current > end || current <= start)
+				break;
+		} else {
+			if (current > end && current < start)
+				break;
+		}
+	}
+}
diff --git a/include/delay.h b/include/delay.h
index c9c5225..070172a 100644
--- a/include/delay.h
+++ b/include/delay.h
@@ -10,20 +10,7 @@
 
 #include "soc.h"
 
-static inline void udelay(unsigned long usec)
-{
-	unsigned long loops = CPU_FREQ_US * usec;
-
-	__asm__ __volatile__(
-	"	mov lp_count, %0	\n"
-	"	lp  1f			\n"
-	"	nop			\n"
-	"1:				\n"
-	:
-        : "r"(loops)
-        : "lp_count");
-}
-
+void udelay(unsigned int usec);
 void arcboot_rom_init_ts(void);
 bool arcboot_rom_timeout_expired(uint64_t start_tv, uint32_t delay);
 uint64_t arcboot_rom_get_ts_us(void);
